src/dtgtk/thumbnail.c

```diff
@@ -646,16 +646,20 @@ static gboolean _event_image_draw(GtkWidget *widget,
       // better use the preview buffer for surface, in order to stay in sync
       dt_thumbnail_surface_destroy(thumb);
 
+      // as we have preview data available we make sure this is processed
+      // while the preview pipe is locked to protect againt darkroom image switching
+      // to next.
+      dt_pthread_mutex_t *mutex = &dev->preview_pipe->backbuf_mutex;
+      dt_pthread_mutex_lock(mutex);
+
       // get new surface with preview image
       const int buf_width = dev->preview_pipe->backbuf_width;
       const int buf_height = dev->preview_pipe->backbuf_height;
-      uint8_t *rgbbuf = g_malloc0(sizeof(unsigned char) * 4 * buf_width * buf_height);
 
-      dt_pthread_mutex_t *mutex = &dev->preview_pipe->backbuf_mutex;
-      dt_pthread_mutex_lock(mutex);
-      memcpy(rgbbuf, dev->preview_pipe->backbuf,
-             sizeof(unsigned char) * 4 * buf_width * buf_height);
-      dt_pthread_mutex_unlock(mutex);
+      const size_t bbufsize = sizeof(uint8_t) * 4 * buf_width * buf_height;
+      uint8_t *rgbbuf = dt_alloc_align_uint8(bbufsize);
+
+      memcpy(rgbbuf, dev->preview_pipe->backbuf, bbufsize);
```

Comment:
It's safe to unlock the mutex immediately after the memcpy, since no more references are made to the pipe's data, only local copies.  I've just checked, and the only place that points `backbuf` at a new buffer also sets `backbuf_width` and `backbuf_height` in the same critical section, so there's no possibility of seeing inconsistent data here.

Reply:
Right, cross checked too.